### PR TITLE
fix sometimes returns response body to HEAD requests

### PR DESCRIPTION
**Pull request by zyfjeff**
Signed-off-by: tianqian.zyf <tianqian.zyf@alibaba-inc.com>

*Description*:  fix sometimes returns response body to HEAD requests
*Risk Level*: low
*Testing*: unittests and integration test
*Docs Changes*:
*Release Notes*:
[Optional Fixes #Issue]  #3932
[Optional *Deprecated*:]

### CODE REVIEW

File: include/envoy/http/filter.h [hunk 1]
```diff
@@ -203,7 +203,8 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
    *                       response headers.
    */
   virtual void sendLocalReply(Code response_code, const std::string& body_text,
-                              std::function<void(HeaderMap& headers)> modify_headers) PURE;
+                              std::function<void(HeaderMap& headers)> modify_headers,
+                              bool is_head_request) PURE;
```
[the remark this answers is not in the document]
Comment (zyfjeff) [reply include/envoy/http/filter.h]: You are right, I will refactor

File: source/common/http/async_client_impl.cc [hunk 2]
```diff
@@ -112,6 +112,10 @@ void AsyncStreamImpl::encodeTrailers(HeaderMapPtr&& trailers) {
 }
 
 void AsyncStreamImpl::sendHeaders(HeaderMap& headers, bool end_stream) {
+  if (headers.Method() &&
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: Method should always be set so I don't think we need to check

File: source/common/http/conn_manager_impl.cc [hunk 3]
```diff
@@ -481,6 +481,11 @@ const Network::Connection* ConnectionManagerImpl::ActiveStream::connection() {
 
 void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
   request_headers_ = std::move(headers);
+  if (request_headers_->Method() &&
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc]: Method can't be null

File: source/common/http/conn_manager_impl.h [hunk 4]
```diff
@@ -188,6 +188,10 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     // called here may change the content type, so we must check it before the call.
     FilterHeadersStatus decodeHeaders(HeaderMap& headers, bool end_stream) {
       is_grpc_request_ = Grpc::Common::hasGrpcContentType(headers);
+      if (headers.Method() &&
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: Method can't be null

File: source/common/http/conn_manager_impl.h [hunk 5]
```diff
@@ -396,6 +402,7 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     // By default, we will assume there are no 100-Continue headers. If encode100ContinueHeaders
     // is ever called, this is set to true so commonContinue resumes processing the 100-Continue.
     bool has_continue_headers_{};
+    bool is_head_request_{false};
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: do we need this at both levels for early response before we have filters?

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 6]
```diff
@@ -75,6 +75,7 @@ Http::FilterHeadersStatus FaultFilter::decodeHeaders(Http::HeaderMap& headers, b
   // faults. In other words, runtime is supported only when faults are
   // configured at the filter level.
   fault_settings_ = config_->settings();
+
```
Comment (mattklein123) [inline source/extensions/filters/http/fault/fault_filter.cc]: revert unrelated change

File: source/extensions/filters/http/grpc_web/grpc_web_filter.cc [hunk 7]
```diff
@@ -45,7 +45,6 @@ Http::FilterHeadersStatus GrpcWebFilter::decodeHeaders(Http::HeaderMap& headers,
     return Http::FilterHeadersStatus::Continue;
   }
   is_grpc_web_request_ = true;
-
```
Comment (mattklein123) [inline source/extensions/filters/http/grpc_web/grpc_web_filter.cc]: revert unrelated change

File: source/extensions/filters/http/jwt_authn/filter.cc [hunk 8]
```diff
@@ -1,5 +1,6 @@
 #include "extensions/filters/http/jwt_authn/filter.h"
 
+#include "common/http/headers.h"
```
Comment (mattklein123) [inline source/extensions/filters/http/jwt_authn/filter.cc]: revert all changes in file

File: source/extensions/filters/http/jwt_authn/filter.h [hunk 9]
```diff
@@ -43,6 +43,7 @@ class Filter : public Http::StreamDecoderFilter,
   State state_ = Init;
   // Mark if request has been stopped.
   bool stopped_ = false;
+  bool is_head_request_{false};
```
Comment (mattklein123) [inline source/extensions/filters/http/jwt_authn/filter.h]: revert

File: source/extensions/filters/http/rbac/rbac_filter.cc [hunk 10]
```diff
@@ -1,5 +1,6 @@
 #include "extensions/filters/http/rbac/rbac_filter.h"
 
+#include "common/http/headers.h"
```
Comment (mattklein123) [inline source/extensions/filters/http/rbac/rbac_filter.cc]: revert

File: source/common/http/conn_manager_impl.h (hunk 5 above)
Comment (zyfjeff) [reply source/common/http/conn_manager_impl.h]: A head request only consider early response this scenario, with filters is upstream service is needed to deal with

File: source/extensions/filters/http/grpc_web/grpc_web_filter.cc (hunk 7 above)
Comment (zyfjeff) [reply source/extensions/filters/http/grpc_web/grpc_web_filter.cc]: Sorry, forgot to check before submission

File: source/extensions/filters/http/fault/fault_filter.cc (hunk 6 above)
Comment (zyfjeff) [reply source/extensions/filters/http/fault/fault_filter.cc]: Sorry, forgot to check before submission

**Comment by zyfjeff** [comment]
`AsyncClient::SendLocalReply` were not in use, I don't know how to write tests, If you know, to help give me some advice.

**Comment by zyfjeff** [comment]
@mattklein123 conn_manager_impl call sendLocalReply is head response, the body is empty, the other HTTP filter test conn_manager_impl didn't go, I don't know how to test, some HTTP filter integration testing, walked conn_manager_impl, I already inside the case，The AsyncClient::sendLocalReply I didn't find a place to call

**Comment by mattklein123** [comment]
> The AsyncClient::sendLocalReply I didn't find a place to call

For AsyncClient, sendLocalReply() can be called in certain error cases where the router does a direct response. For example in this test (https://github.com/envoyproxy/envoy/blob/master/test/common/http/async_client_impl_test.cc#L466) I think if you change that to a HEAD request you will hit the code you modified.

> @mattklein123 conn_manager_impl call sendLocalReply is head response, the body is empty, the other HTTP filter test conn_manager_impl didn't go, I don't know how to test, some HTTP filter integration testing, walked conn_manager_impl, I already inside the test case

It would be nice to have a conn_manager_impl_test for this, though nice job adding the integration tests, so I'm not going to block on it, but if you feel like it would be great. Check out https://github.com/envoyproxy/envoy/blob/master/test/common/http/conn_manager_impl_test.cc#L2880 and the tests around it. You should be able to setup a test like that, and do a local reply from within various filter callbacks which should drive the various flows.

Thank you!

**Comment by zyfjeff** [comment]
@mattklein123  AsyncClient and conn_manager_impl added some unittests

File: test/common/http/utility_test.cc [hunk 11]
```diff
@@ -359,7 +359,18 @@ TEST(HttpUtility, SendLocalReplyDestroyedEarly) {
     is_reset = true;
   }));
   EXPECT_CALL(callbacks, encodeData(_, true)).Times(0);
-  Utility::sendLocalReply(false, callbacks, is_reset, Http::Code::PayloadTooLarge, "large");
+  Utility::sendLocalReply(false, callbacks, is_reset, Http::Code::PayloadTooLarge, "large", false);
+}
+
+TEST(HttpUtility, SendLocalReplayHeaderReply) {
```
Comment (mattklein123) [inline test/common/http/utility_test.cc]: typo "Replay". I think maybe you mean "SendLocalReplyHeadRequest" ?

File: source/common/http/utility.h [hunk 12]
```diff
@@ -131,9 +131,10 @@ Http1Settings parseHttp1Settings(const envoy::api::v2::core::Http1ProtocolOption
  * @param response_code supplies the HTTP response code.
  * @param body_text supplies the optional body text which is sent using the text/plain content
  *                  type.
+ * @param is_head_request tells if this is a HEAD request
```
Comment (alyssawilk) [inline source/common/http/utility.h]: if this is a response to a HEAD request?

**Comment by mattklein123** [comment]
> As always I'm a big fan of integration tests

There are 2 integration tests in the change (JWT filter and RBAC filter)

**Comment by alyssawilk** [comment]
Sorry, I could have said "we could also copy-paste" as I'd mildly prefer having an integration test which does not depend on filters.    The one I suggested validates the behavior in the http connection manager for all 4 protocol combinations and permutations, which I believe has value.

**Comment by mattklein123** [comment]
Ah yes, that would be good, agreed.

File: test/common/http/utility_test.cc (hunk 11 above)
Comment (zyfjeff) [reply test/common/http/utility_test.cc]: yes, spelling mistakes

File: source/common/http/utility.h (hunk 12 above)
Comment (zyfjeff) [reply source/common/http/utility.h]: Yes, I fix the syntax mistakes

**Comment by zyfjeff** [comment]
@mattklein123  I met some problems when doing the integration test, has not yet been solved, can you help a look.

1. I started a Head request
```
  IntegrationStreamDecoderPtr setupPerStreamIdleTimeoutHeadRequestTest() {
    initialize();
    fake_upstreams_[0]->set_allow_unexpected_disconnects(true);
    codec_client_ = makeHttpConnection(makeClientConnection((lookupPort("http"))));
    auto encoder_decoder =
        codec_client_->startRequest(Http::TestHeaderMapImpl{{":method", "HEAD"},
                                                            {":path", "/test/long/url"},
                                                            {":scheme", "http"},
                                                            {":authority", "host"}});
    request_encoder_ = &encoder_decoder.first;
    auto response = std::move(encoder_decoder.second);
    fake_upstream_connection_ = fake_upstreams_[0]->waitForHttpConnection(*dispatcher_);
    upstream_request_ = fake_upstream_connection_->waitForNewStream(*dispatcher_);
    upstream_request_->waitForHeadersComplete();
    return response;
  }
```

2. Whether the head only request or other types of request end_stream is false
```
std::pair<Http::StreamEncoder&, IntegrationStreamDecoderPtr>
IntegrationCodecClient::startRequest(const Http::HeaderMap& headers) {
  auto response = std::make_unique<IntegrationStreamDecoder>(dispatcher_);
  Http::StreamEncoder& encoder = newStream(*response);
  encoder.getStream().addCallbacks(*response);
  // Here whether the head only request or with data, is false
  encoder.encodeHeaders(headers, false);
  flushWrite();
  return {encoder, std::move(response)};
}
```

3. This leads to the codec for HEAD request there is no set its response to the HEAD rsponse
```
void StreamEncoderImpl::encodeData(Buffer::Instance& data, bool end_stream) {
  // end_stream may be indicated with a zero length data buffer. If that is the case, so not
  // atually write the zero length buffer out.
  if (data.length() > 0) {
    if (chunk_encoding_) {
      connection_.buffer().add(fmt::format("{:x}\r\n", data.length()));
    }

    connection_.buffer().move(data);

    if (chunk_encoding_) {
      connection_.buffer().add(CRLF);
    }
  }

  if (end_stream) {
    endEncode();   // Don't perform endEncode because endstream is false
  } else {
    connection_.flushOutput();
  }
}

void StreamEncoderImpl::endEncode() {
  if (chunk_encoding_) {
    connection_.buffer().add(LAST_CHUNK);
  }

  connection_.flushOutput();
  connection_.onEncodeComplete();
}

void ClientConnectionImpl::onEncodeComplete() {
  // Transfer head request state into the pending response before we reuse the encoder.
  pending_responses_.back().head_request_ = request_encoder_->headRequest();
}
```
4. The final test to find the response->compete() always returns false

```
[2018-08-06 11:23:15.466][16169][trace][http] source/common/http/http1/codec_impl.cc:341] [C48] parsing 0 bytes
[2018-08-06 11:23:15.466][16169][info][client] source/common/http/codec_client.cc:117] [C48] protocol error: http/1.1 protocol error: HPE_INVALID_EOF_STATE
[2018-08-06 11:23:15.466][16169][debug][client] source/common/http/codec_client.cc:81] [C48] disconnect. resetting 1 pending requests
[2018-08-06 11:23:15.466][16169][debug][client] source/common/http/codec_client.cc:104] [C48] request reset
[2018-08-06 11:23:15.466][16169][trace][main] source/common/event/dispatcher_impl.cc:126] item added to deferred deletion list (size=1)
[2018-08-06 11:23:15.466][16169][trace][main] source/common/event/dispatcher_impl.cc:52] clearing deferred deletion list (size=1)
test/integration/idle_timeout_integration_test.cc:104: Failure
Value of: response->complete()
  Actual: false
Expected: true
```

5. I repaired by changing the code

```
std::pair<Http::StreamEncoder&, IntegrationStreamDecoderPtr>
IntegrationCodecClient::startRequest(const Http::HeaderMap& headers) {
  auto response = std::make_unique<IntegrationStreamDecoder>(dispatcher_);
  Http::StreamEncoder& encoder = newStream(*response);
  encoder.getStream().addCallbacks(*response);
 // Determine whether it is a HEAD request to set the value of end_stream
  bool end_stream = (headers.Method()->value().c_str() == Http::Headers::get().MethodValues.Head);
  encoder.encodeHeaders(headers, end_stream);
  flushWrite();
  return {encoder, std::move(response)};
}
```

6.  Test found that waitForDisconnect can not return

```
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:341] [C48] parsing 130 bytes
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:439] [C48] message begin
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:309] [C48] completed header: key=content-length value=14
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:309] [C48] completed header: key=content-type value=text/plain
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:309] [C48] completed header: key=date value=Mon, 06 Aug 2018 11:28:09 GMT
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:405] [C48] headers complete
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:309] [C48] completed header: key=server value=envoy
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:426] [C48] message complete
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:729] [C48] message complete
[2018-08-06 11:28:09.849][22301][debug][client] source/common/http/codec_client.cc:94] [C48] response complete
[2018-08-06 11:28:09.849][22301][trace][main] source/common/event/dispatcher_impl.cc:126] item added to deferred deletion list (size=1)
[2018-08-06 11:28:09.849][22301][trace][http] source/common/http/http1/codec_impl.cc:362] [C48] parsed 130 bytes
[2018-08-06 11:28:09.849][22301][trace][main] source/common/event/dispatcher_impl.cc:52] clearing deferred deletion list (size=1)
[2018-08-06 11:28:09.849][22443][trace][connection] source/common/network/raw_buffer_socket.cc:67] [C49] write returns: 130
[2018-08-06 11:28:14.638][22438][debug][main] source/server/server.cc:119] flushing stats
[2018-08-06 11:28:19.639][22438][debug][main] source/server/server.cc:119] flushing stats
[2018-08-06 11:28:24.639][22438][debug][main] source/server/server.cc:119] flushing stats
[2018-08-06 11:28:29.639][22438][debug][main] source/server/server.cc:119] flushing stats
[2018-08-06 11:28:34.640][22438][debug][main] source/server/server.cc:119] flushing stats
```

**Comment by alyssawilk** [comment]
If you stick the whole test log with -l trace to pastebin I'll take a look.  If there's no disconnect it's possible Envoy thinks it has handled the full request/response pair.  Given startRequest I'd think it should be in "handling request" when the idle timeout comes and so end the connection.  Given we (unfortunately) allow GET with body I assume we allow HEAD with body, but if I can see full logs I might be able to help you puzzle out what's going on.

**Comment by zyfjeff** [comment]
@alyssawilk [test.log](https://paste.ubuntu.com/p/S69FgqzDk3/), Connection should be closed, I didn't see related connection through the netstat command information, through the log view, seems to be caused by connection was closed early.  

I modify `IntegrationCodecClient: : startRequest` of `encoder.EncodeHeaders (headers, false)` To the `encoder EncodeHeaders (headers, true)`  in latest master code, Also appear the same problem, log as follows
[test.log.2](https://paste.ubuntu.com/p/J6BP6Tw6Rr/)

Modify the diff content is as follows:
```
diff --git a/test/integration/http_integration.cc b/test/integration/http_integration.cc
index 3f93b90..c756b49 100644
--- a/test/integration/http_integration.cc
+++ b/test/integration/http_integration.cc
@@ -137,7 +137,7 @@ IntegrationCodecClient::startRequest(const Http::HeaderMap& headers) {
   auto response = std::make_unique<IntegrationStreamDecoder>(dispatcher_);
   Http::StreamEncoder& encoder = newStream(*response);
   encoder.getStream().addCallbacks(*response);
-  encoder.encodeHeaders(headers, false);
+  encoder.encodeHeaders(headers, true);
   flushWrite();
   return {encoder, std::move(response)};
 }
```

Thank you very much can you help me to look at this problem

**Comment by alyssawilk** [comment]
Ah, see if you
+  encoder.encodeHeaders(headers, true);
then you've sent a full request.

When we send the idle response, we've received a full request* and sent a full (idle timeout) response, then Envoy won't disconnect for HTTP/1 - it'll reuse the connection for a subsequent response

If you want to use waitForTimeout, you have to have the timeout occur during request upload, so have an incomplete request.  It's also fine by me if you don't do an incomplete request, and simply wait for a full response without the disconnect.

*[2018-08-06 15:59:11.424][29144][trace][http] source/common/http/http1/codec_impl.cc:426] [C49] message complete

**Comment by zyfjeff** [comment]
@alyssawilk  if `encoder.encodeHeaders(headers, true)`, lead to  can't transfor head request state  into the pending response. 

For incomplete request `StreamEncoderImpl::encodeHeaders` seems to be problematic
```
void StreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_stream) {
.......

  if (end_stream) {
    endEncode();
  } else {
    connection_.flushOutput();
  }
}
```
I tried to fix the problem, help me review the code and see whether it is right

File: source/common/http/http1/codec_impl.cc [hunk 13]
```diff
@@ -119,6 +119,11 @@ void StreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_stream)
   if (end_stream) {
     endEncode();
   } else {
+    // Request is likely to be incomplete, in this case need to transfer the head request state into
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: Sorry, I'm not clear on why this is necessary.  Our existing HEAD processing correctly determines when the headers are processed that the encode is complete and the new code path using sendLocalReply sets end stream explicitly.  Why do we need to do it manually in this new code path?

Comment (zyfjeff) [reply source/common/http/http1/codec_impl.cc]: If  this is an  incomplete request (end_stream=false), then `connection_.onEncodeComplete();` will not be called,  the head request state will not be passed to the corresponding response, causing http_parser to parse the response. "protocol error: HTTP / 1.1 protocol error: HPE_INVALID_EOF_STATE" appears, there is no way to make "expect true" for response->complete() in IdleTimeoutIntegrationTest

**Comment by zyfjeff** [comment]
@alyssawilk  I'm sorry to trouble you, i will read the contributors guidelines carefully.

File: source/common/http/http1/codec_impl.cc (hunk 13 above)
Comment (alyssawilk) [reply source/common/http/http1/codec_impl.cc]: Gotcha.   I think what this functionally does though is disallow HEAD requests with bodies.  I'd be very happy to disallow HEAD requests with bodies, but if we're going to do that I believe we should disallow GETs with bodies.  Either I think would be a higher risk change and I think should land as a separate PR :-)

If the problem is getting the information of "this is a response to the head request" to the response earlier, could we call
'pending_responses_.back().head_request_ = request_encoder_->headRequest();`
earlier in the chain, say when we've received the request headers?   @mattklein123 for attention on this one since it involves codec internals fairly deeply.

Comment (mattklein123) [reply source/common/http/http1/codec_impl.cc]: > Gotcha. I think what this functionally does though is disallow HEAD requests with bodies. I'd be very happy to disallow HEAD requests with bodies, but if we're going to do that I believe we should disallow GETs with bodies. Either I think would be a higher risk change and I think should land as a separate PR :-)

Does it? I'm not sure it does?

> If the problem is getting the information of "this is a response to the head request" to the response earlier, could we call
'pending_responses_.back().head_request_ = request_encoder_->headRequest();`
earlier in the chain, say when we've received the request headers? @mattklein123 for attention on this one since it involves codec internals fairly deeply.

Yes, +1, we should do this outside of this function, and optimally only on the client side when headers are encoded. I think the optimal thing to do here would be to make a `ClientStreamEncoder` and override encodeHeaders() and then do this at the beginning of the call, before calling the base class encodeHeaders(). This is a larger change but I think would be more clear. Thoughts?

File: source/common/http/http1/codec_impl.cc [hunk 14]
```diff
@@ -687,7 +692,9 @@ StreamEncoder& ClientConnectionImpl::newStream(StreamDecoder& response_decoder)
 
 void ClientConnectionImpl::onEncodeComplete() {
   // Transfer head request state into the pending response before we reuse the encoder.
-  pending_responses_.back().head_request_ = request_encoder_->headRequest();
+  if (!pending_responses_.empty()) {
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: Why is this change needed? Because of the other change? I'm hoping by doing what I suggest in the other change we can avoid this?

Comment (zyfjeff) [reply source/common/http/http1/codec_impl.cc]: Since `RequestStreamEncoderImpl` is called in ws_handler_impl, it internally calls`StreamEncoderImpl::encodeHeaders`

```
Http1::RequestStreamEncoderImpl upstream_request = Http1::RequestStreamEncoderImpl(upstream_http);
```
This` RequestStreamEncoderImpl` is not created by `ClientConnectionImpl::newStream`. At this time, the peningding_response inside ClientConnectionImpl does not have any elements, so calling `pending_responses_.back().head_request_ = request_encoder_->headRequest();` will cause problems.

```
StreamEncoder& ClientConnectionImpl::newStream(StreamDecoder& response_decoder) {
  if (resetStreamCalled()) {
    throw CodecClientException("cannot create new streams after calling reset");
  }
  // Streams are responsible for unwinding any outstanding readDisable(true)
  // calls done on the underlying connection as they are destroyed. As this is
  // the only place a HTTP/1 stream is destroyed where the Network::Connection is
  // reused, unwind any outstanding readDisable() calls here.
  while (!connection_.readEnabled()) {
    connection_.readDisable(false);
  }
  request_encoder_.reset(new RequestStreamEncoderImpl(*this));
  pending_responses_.emplace_back(&response_decoder);
  return *request_encoder_;
}
```

**Comment by zyfjeff** [comment]
@mattklein123 @alyssawilk I agree with your view, I re-adjusted the call point of connection_.onEncodeComplete() and put it in the place where the request header was received.

File: source/common/http/http1/codec_impl.cc (hunk 13 above)
Comment (alyssawilk) [reply source/common/http/http1/codec_impl.cc]: Yeah, that'd work for me.   Looking at onEncodeComplete the only thing it does is push the state of head request, so rather than call onEncodeComplete when we're not done encodnig (end stream == false) I'm all for doing it when we get the headers.

Thanks for bearing with us, @zyfjeff while we sort this out :-)

File: source/common/http/http1/codec_impl.cc [hunk 15]
```diff
@@ -247,6 +247,11 @@ void RequestStreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_
 
   if (method->value() == Headers::get().MethodValues.Head.c_str()) {
     head_request_ = true;
+    // If this is an incomplete request (end_stream=false), then connection_.onEncodeComplete();
+    // will not be called in StreamEncoderImpl::encodeHeaders, head request state will not be passed
+    // to the corresponding response so here need to call this method
+    if (!end_stream)
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: nit: needs braces around this if statement. 

I'm still not a fan of calling `onEncodeComplete` here as the encode is not necessarily complete. Can we just pass the state directly? This is what I was referring to previously. Can the the RequestStreamEncoderImpl hold a reference to the ClientConnection such that we can transfer this state directly? I think that might be cleaner?

Comment (zyfjeff) [reply source/common/http/http1/codec_impl.cc]: You actually think that the name of onEncodeComplete will have some ambiguity. In fact, there is no complete code completion. 

Can I directly dynamic_cast from ConnectionImpl to ClientConnectionImpl and pass the state directly?

File: source/common/http/http1/codec_impl.cc [hunk 16]
```diff
@@ -247,6 +247,15 @@ void RequestStreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_
 
   if (method->value() == Headers::get().MethodValues.Head.c_str()) {
     head_request_ = true;
+    // If this is an incomplete request (end_stream=false), then connection_.onEncodeComplete();
+    // will not be called in StreamEncoderImpl::encodeHeaders, head request state will not be passed
+    // to the corresponding response so here need to call this method
+    if (!end_stream) {
+      auto c = dynamic_cast<ClientConnectionImpl*>(&connection_);
+      if (c) {
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: This should always be true. If we are going to do the dynamic cast we can just assume it is valid, though I'm not a huge fan of dynamic_cast for something like this. The other option is to store `ClientConncetion` in `RequestStreamEncoderImpl` (and server connection in `ResponseStreamEncoderImpl` and then have a base class virtual function `connection()` which returns the connection base class. @alyssawilk any opinions here?

Comment (alyssawilk) [reply source/common/http/http1/codec_impl.cc]: I'd suggest that just like we have a virtual onEncodeComplete that the client/server connections implement, we could have a pure virtual onEncodeHeaders.  We could handle passing on head status in the ClientConnectionImpl and the ServerConnectionImpl could just be a no-op for now.

I think that would be consistent with the style of the rest of the code base, and make it clear what part of the request/response pipeline is being handled.

File: source/common/http/http1/codec_impl.cc [hunk 17]
```diff
@@ -685,9 +684,12 @@ StreamEncoder& ClientConnectionImpl::newStream(StreamDecoder& response_decoder)
   return *request_encoder_;
 }
 
-void ClientConnectionImpl::onEncodeComplete() {
-  // Transfer head request state into the pending response before we reuse the encoder.
-  pending_responses_.back().head_request_ = request_encoder_->headRequest();
+void ClientConnectionImpl::onEncodeComplete() {}
+
+void ClientConnectionImpl::onEncodeHeader(const HeaderMapImpl& headers) {
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/http/http1/codec_impl.cc]: Sorry, I think there was a misunderstanding on this one.
I was asking the name go from onEncodeHeader -> onEncodeHeaders (add an "s")
We don't need the same arguments, just Header with an S :-)
If you can remove the extra argument and change the name that would be great!

**Comment by zyfjeff** [comment]
@alyssawilk  @mattklein123  Thank you for some review suggestions.

File: source/common/http/http1/codec_impl.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/http/http1/codec_impl.cc]: LGTM once this comment is addressed :-)

**Comment by zyfjeff** [comment]
@alyssawilk If this PR wants to merge, what else do I need to do?

File: source/common/http/http1/codec_impl.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (zyfjeff) [reply source/common/http/http1/codec_impl.cc]: Sorry,  I have never seen this comment (onEncodeHeader -> onEncodeHeaders (add an "s"), so I have not paid attention to this PR in recent days.
